src/php/ext/grpc/channel.c

```diff
@@ -146,14 +180,93 @@ PHP_METHOD(Channel, __construct) {
+    }
+    php_grpc_zend_hash_del(array_hash, "force_new", sizeof("force_new"));
+  }
+
+  // parse the rest of the channel args array
+  if (php_grpc_read_args_array(args_array, &args TSRMLS_CC) == FAILURE) {
+    return;
+  }
+
+  // Construct a hashkey for the persistent channel
+  // Currently, the hashkey contains 3 parts:
+  // 1. hostname
+  // 2. hash value of the channel args array (excluding "credentials"
+  //    and "force_new")
+  // 3. (optional) hash value of the ChannelCredentials object
+  //
+  // Note that, if the ChannelCredentials object was created by
+  // "composing" a ChannelCredentials and CallCredentials object
+  // together, the CallCredentials portion will be ignored, when
+  // constructing this hashkey.
+  //
+  // In other words, the same ChannelCredentials object, composed
+  // with 2 different CallCredentials objects, will return the same
+  // underlying grpc_channel. So proceed with caution.
+
+  php_serialize_data_t var_hash;
+  smart_str buf = {0};
+  PHP_VAR_SERIALIZE_INIT(var_hash);
+  PHP_GRPC_VAR_SERIALIZE(&buf, args_array, &var_hash);
+  PHP_VAR_SERIALIZE_DESTROY(var_hash);
+
+  char sha1str[41];
+  generate_sha1_str(sha1str, PHP_GRPC_SERIALIZED_BUF_STR(buf),
+                    PHP_GRPC_SERIALIZED_BUF_LEN(buf));
+
+  php_grpc_int key_len = target_length + strlen(sha1str);
+  if (creds != NULL && creds->hashstr != NULL) {
+    key_len += strlen(creds->hashstr);
+  }
+  char *key = malloc(key_len + 1);
+  strcpy(key, target);
+  strcat(key, sha1str);
+  if (creds != NULL && creds->hashstr != NULL) {
+    strcat(key, creds->hashstr);
+  }
+  channel->wrapper = malloc(sizeof(grpc_channel_wrapper));
+  channel->wrapper->key = key;
+  gpr_mu_init(&channel->wrapper->mu);
+  smart_str_free(&buf);
+
+  if (force_new) {
+    php_grpc_delete_persistent_list_entry(key, key_len TSRMLS_CC);
+  }
+
+  if (!(PHP_GRPC_PERSISTENT_LIST_FIND(&EG(persistent_list), key, key_len,
+                                      rsrc))) {
+    php_grpc_zend_resource new_rsrc;
+    // this links each persistent list entry to a destructor
+    new_rsrc.type = le_plink;
+    le = malloc(sizeof(channel_persistent_le_t));
+
+    if (creds == NULL) {
+      channel->wrapper->wrapped = grpc_insecure_channel_create(target, &args,
+                                                               NULL);
+    } else {
+      channel->wrapper->wrapped =
+          grpc_secure_channel_create(creds->wrapped, target, &args, NULL);
+    }
+    efree(args.args);
+    le->channel = channel->wrapper;
+    new_rsrc.ptr = le;
+    PHP_GRPC_PERSISTENT_LIST_UPDATE(&EG(persistent_list), key, key_len,
+                                    (void *)&new_rsrc);
   } else {
-    channel->wrapped =
-        grpc_secure_channel_create(creds->wrapped, target, &args, NULL);
+    le = (channel_persistent_le_t *)rsrc->ptr;
+    if (strcmp(target, grpc_channel_get_target(le->channel->wrapped)) != 0) {
+      zend_throw_exception(spl_ce_RuntimeException,
```

Comment:
I think we should just evict the old channel from the cache, and replace it with a new channel.  We don't want a cache collision to kill an application.

Also, can we just dump all the fields used for the hash in the new ```wrapper``` object?  This way we can actually verify we have the same channel, not just the same target.